qutip/solver/result.py

```diff
@@ -861,19 +853,166 @@ def __add__(self, other):
                 new.e_data = new.runs_e_data
 
         new.stats["run time"] += other.stats["run time"]
-        new.stats['end_condition'] = "Merged results"
+        new.stats["end_condition"] = "Merged results"
 
         return new
 
 
+class MultiTrajResultImprovedSampling(MultiTrajResult):
+    """
+    See docstring for MultiTrajResult for all relevant documentation.
+    This child class computes expectation values and sums of states, etc
+    using the improved sampling algorithm, which samples the no-jump trajectory
+    first and then only samples jump trajectories afterwards.
+    """
+    def __init__(self, e_ops, options, **kw):
+        super().__init__(e_ops, options, **kw)
+        self._sum_expect_no_jump = None
+        self._sum_expect_jump = None
+        self._sum2_expect_no_jump = None
+        self._sum2_expect_jump = None
+
+        self._sum_states_no_jump = None
+        self._sum_states_jump = None
+        self._sum_final_states_no_jump = None
+        self._sum_final_states_jump = None
+
+        self.no_jump_prob = None
+
+    def _store_trajectory(self, trajectory, no_jump=False):
+        super()._store_trajectory(trajectory)
+
+    def _reduce_states(self, trajectory, no_jump=False):
+        if no_jump:
+            self._sum_states_no_jump = [
+                accu + self._to_dm(state)
+                for accu, state in zip(self._sum_states_no_jump, trajectory.states)
+            ]
+        else:
+            self._sum_states_jump = [
+                accu + self._to_dm(state)
+                for accu, state in zip(self._sum_states_jump, trajectory.states)
+            ]
+
+    def _increment_traj(self, trajectory, no_jump=False):
+        super()._increment_traj(trajectory)
+
+    def _reduce_final_state(self, trajectory, no_jump=False):
+        dm_final_state = self._to_dm(trajectory.final_state)
+        if no_jump:
+            self._sum_final_states_no_jump += dm_final_state
+        else:
+            self._sum_final_states_jump += dm_final_state
+
+    def _average_computer(self):
+        avg = np.array(self._sum_expect_jump) / (self.num_trajectories - 1)
+        avg2 = np.array(self._sum2_expect_jump) / (self.num_trajectories - 1)
+        return avg, avg2
+
+    def _add_first_traj(self, trajectory):
```

Comment:
Agreed re testing num_trajectories!=1, that's more elegant and would eliminate the need for a custom `add` function. Where did you have in mind when you say "here"?

Reply:
By here, I meant that the logic to add the no jump trajectory could be in `_add_first_traj`.

Reply:
I'm not positive that would be the best option, since the purpose of `_add_first_traj` is only to initialize a bunch of zero arrays (could be a motivation to change the name of that function to something like `_initialize_data_arrays`). All of the data for each trajectory, including the first trajectory, is added by the `_reduce*` functions. 

I could add the no-jump trajectory at the end of `_add_first_traj` but I would still have to check e.g. `if self.num_trajectories != 1:` in each of the `_reduce*` functions. For now I've committed a version that leaves the behavior of `_add_first_traj` unmodified

Reply:
I'm not positive that would be the best option, since the purpose of `_add_first_traj` is only to initialize a bunch of zero arrays (could be a motivation to change the name of that function to something like `_initialize_data_arrays`). All of the data for each trajectory, including the first trajectory, is added by the `_reduce*` functions.

I could add the no-jump trajectory at the end of `_add_first_traj` but I would still have to check e.g. `if self.num_trajectories != 1:` in each of the `_reduce*` functions. For now I've committed a version that leaves the behavior of `_add_first_traj` unmodified